tests/InferTypesTest.elm

```diff
@@ -8,47 +8,142 @@ import Common
+              )
+            , ( "different types"
+              , Canonical.List [ Canonical.Literal (Int 1), Canonical.Literal (String "two") ]
+              , Err (Error.TypeMismatch Type.Int Type.String)
               )
+            , ( "more items with different types"
+              , Canonical.List [ Canonical.Literal (Bool True), Canonical.Literal (String "two"), Canonical.Literal (Int 3) ]
+              , Err (Error.TypeMismatch Type.Bool Type.String)
+              )
+            ]
+        , describe "fuzz exprInfer"
+            [ fuzzExpr Type.Int
+            , fuzzExpr Type.Float
+            , fuzzExpr Type.Bool
+            , fuzzExpr Type.Char
+            , fuzzExpr Type.String
+            , fuzzExpr Type.Unit
             ]
-        )
+        ]
+
+
+runSection : String -> List ( String, Canonical.Expr, Result Error.TypeError Typed.Expr ) -> Test
+runSection description tests =
+    describe description
+        (List.map runTest tests)
+
+
+runTest : ( String, Canonical.Expr, Result Error.TypeError Typed.Expr ) -> Test
+runTest ( description, input, output ) =
+    test description <|
+        \() ->
+            Stage.InferTypes.inferExpr input
+                |> Expect.equal output
+
+
+fuzzExpr : Type -> Test
+fuzzExpr typeWanted =
+    let
+        description =
+            Type.toString typeWanted
+    in
+    fuzz (randomExprFromType typeWanted) description <|
+        \input ->
+            Stage.InferTypes.inferExpr input
+                |> Result.map Tuple.second
+                |> Expect.equal (Ok typeWanted)
+
+
+randomExprFromType : Type -> Fuzzer Canonical.Expr
+randomExprFromType targetType =
+    let
+        cannotFuzz () =
+            Debug.todo <| "Cannot fuzz " ++ Type.toString targetType ++ " expressions."
+    in
+    case targetType of
+        Type.Int ->
+            intExpr
+
+        Type.Float ->
+            floatExpr
+
+        Type.Bool ->
+            boolExpr
+
+        Type.Char ->
+            charExpr
+
+        Type.String ->
+            stringExpr
+
+        Type.Unit ->
+            unitExpr
+
+        _ ->
+            cannotFuzz ()
+
+
+intExpr : Fuzzer Canonical.Expr
+intExpr =
+    Fuzz.int
+        |> Fuzz.map Int
+        |> Fuzz.map Canonical.Literal
+
+
+floatExpr : Fuzzer Canonical.Expr
+floatExpr =
+    Fuzz.float
+        |> Fuzz.map Float
+        |> Fuzz.map Canonical.Literal
+
+
+boolExpr : Fuzzer Canonical.Expr
+boolExpr =
+    Fuzz.bool
+        |> Fuzz.map Bool
+        |> Fuzz.map Canonical.Literal
+
+
+charExpr : Fuzzer Canonical.Expr
+charExpr =
+    Fuzz.intRange 0 0x0010FFFF
+        |> Fuzz.map Char.fromCode
+        |> Fuzz.map Char
+        |> Fuzz.map Canonical.Literal
+
+
+stringExpr : Fuzzer Canonical.Expr
+stringExpr =
+    Fuzz.intRange 0 0x0010FFFF
+        |> Fuzz.list
+        |> Fuzz.map (List.map Char.fromCode)
+        |> Fuzz.map String.fromList
```

Comment:
That's pretty bold :smile: but I guess for this usecase it doesn't matter if the strings are valid Unicode or not.